crates/pathfinder/src/state/sync/l2.rs

```diff
@@ -201,55 +192,73 @@ pub async fn sync(
     }
 }
 
-/// Download and emit newly declared contract classes.
+/// Download and emit new contract classes.
+///
+/// New classes can come from:
+/// - DECLARE transactions
+/// - `old_declared_contracts` from the state diff (Cairo 0.x classes)
+/// - `declared_classes` from the state diff (Cairo 1.0 classes)
+/// - `deployed_contracts` from the state diff (DEPLOY transactions)
+/// - `replaced_classes` from the state diff
 ///
-/// We cannot remove the older way using `deploy_contracts` as this
-/// is required to handle older blocks which don't have declare transactions.
-async fn declare_classes(
+/// Note that due to an issue with the sequencer previously undeclared classes
+/// can show up in `replaced_classes`. This is caused by DECLARE v0 transactions
+/// that were _failing_ but the sequencer has still added the class to its list of
+/// known classes...
```

Comment:
What is this referring to? Does the gateway recognise failed v0 class declarations and adds the class anyway? That seems really.. broken?

Reply:
It _is_ broken. According to Starkware we have to live with that for now. Basically _anywhere_ we find a class hash that might be a Cairo 0.x class we're missing. We can work around it at a couple of places (like replaced classes) and just ignore it for now in other parts (like execution).

Reply:
Just restating for prosperity:

1. The sequencer has a bug in which it keeps v0 classes that form part of failed transactions
2. These ghost classes are allowed (by the sequencer) to form part of class replacement
3. Such a replaced (ghost) class hash could therefore be the 1st time pathfinder sees this class hash
4. Therefore pathfinder must also download all replaced class hashes as if they were declared